src/main/java/org/roaringbitmap/RunContainer.java

```diff
@@ -942,6 +932,40 @@ public boolean equals(Object o) {
     return false;
   }
 
+  private boolean equals(RunContainer runContainer) {
+    if (runContainer.nbrruns != this.nbrruns) {
+      return false;
+    }
+    for (int i = 0; i < nbrruns; ++i) {
+      if (getValue(i) != runContainer.getValue(i)) {
+        return false;
+      }
+      if (getLength(i) != runContainer.getLength(i)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean equals(ArrayContainer arrayContainer) {
+    int pos = 0;
+    for (short i = 0; i < nbrruns; ++i) {
+      short runStart = getValue(i);
+      short length = getLength(i);
+      if (pos + length >= arrayContainer.getCardinality()) {
+        return false;
```

Comment:
Indeed, the final cardinality check is mandatory now that the initial one was moved in the loop.
Randomised testing is one way to improve your code coverage and the overall quality of your software. In my opinion, it is useful to spot hidden bugs. But I personally prefer to write unit tests first. I've always seen the return on investment worth it.
Note that if the test suite was not failing this morning, it's just because it was not covering all the new possible cases introduced in the PR. And there was no need to cover them since the existing code was handling all cases in a generic way.